Redis Cluster slave "down" on slave connect timeout when startup

<!--
Сonsider Redisson PRO https://redisson.pro version for advanced features and support by SLA.
-->

**Expected behavior**
Redis Cluster slave  reconnect or report error on startup after connect timeout

**Actual behavior**
Redis Cluster slave report "down"

**Steps to reproduce or test case**
1. Cluster slave connect timeout on startup   

```
unable to add slave for: redis://xxxx:xxxx slot ranges: [[9832-13107]] org.redisson.client.RedisConnectionException: Unable to init enough connections amount! Only 14 of 24 were initialized. Redis server: /xxxxxx:xxxx
 	at org.redisson.connection.pool.ConnectionPool$1.lambda$run$0(ConnectionPool.java:160)
        at org.redisson.misc.RedissonPromise.lambda$0(RedissonPromise.java:186)
 	at org.redisson.misc.RedissonPromise.lambda$onComplete$1(RedissonPromise.java:184)
 	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577)
 	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:551)
 	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:490)
 	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:615)
 	at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:608)
 	at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:117)
 	at org.redisson.misc.RedissonPromise.tryFailure(RedissonPromise.java:97)
 	at org.redisson.connection.pool.ConnectionPool.promiseFailure(ConnectionPool.java:332)
 	at org.redisson.connection.pool.ConnectionPool.lambda$createConnection$1(ConnectionPool.java:298)
 	at org.redisson.misc.RedissonPromise.lambda$0(RedissonPromise.java:186)
 	at org.redisson.misc.RedissonPromise.lambda$onComplete$1(RedissonPromise.java:184)
 	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577)
 	at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:570)
 	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:549)
 	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:490)
 	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:615)
 	at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:608)
 	at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:117)
 	at org.redisson.misc.RedissonPromise.tryFailure(RedissonPromise.java:97)
 	at org.redisson.client.RedisClient$2$2.run(RedisClient.java:248)
 	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164) 
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
  [1 skipped]
 Caused by: io.netty.channel.ConnectTimeoutException: connection timed out: /xxxxx:xxxx
 	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(AbstractNioChannel.java:261) 
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98) 
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170) 
... 7 common frames omitted
```

2. cluster slave report error on every command 

```
org.redisson.client.RedisConnectionException: SlaveConnectionPool no available Redis entries. Master entry host: /xxxxx:xxx Disconnected hosts: [/xxxx:xxx]
	at org.redisson.connection.pool.ConnectionPool.get(ConnectionPool.java:219)
	at org.redisson.connection.pool.SlaveConnectionPool.get(SlaveConnectionPool.java:30)
	at org.redisson.connection.balancer.LoadBalancerManager.nextConnection(LoadBalancerManager.java:253)
	at org.redisson.connection.MasterSlaveEntry.connectionReadOp(MasterSlaveEntry.java:499)
	at org.redisson.connection.MasterSlaveConnectionManager.connectionReadOp(MasterSlaveConnectionManager.java:575)
	at org.redisson.command.RedisExecutor.getConnection(RedisExecutor.java:642)
	at org.redisson.command.RedisExecutor.execute(RedisExecutor.java:114)
	at org.redisson.command.CommandAsyncService.async(CommandAsyncService.java:604)
	at org.redisson.command.CommandAsyncService.readAsync(CommandAsyncService.java:367)
	at org.redisson.RedissonBucket.getAsync(RedissonBucket.java:109)
	at org.redisson.RedissonBucket.get(RedissonBucket.java:104)
```

**Redis version**
4.0.12 
**Redisson version**
3.13.4

**Redisson configuration**

```yml

clusterServersConfig:
      idleConnectionTimeout: 10000
      connectTimeout: 100
      timeout: 200
      retryAttempts: 0
      retryInterval: 200
      failedSlaveReconnectionInterval: 1000
      failedSlaveCheckInterval: 1000
      subscriptionsPerConnection: 5
      slaveSubscriptionConnectionMinimumIdleSize: 1
      slaveSubscriptionConnectionPoolSize: 50
      slaveConnectionMinimumIdleSize: 32
      slaveConnectionPoolSize: 64
      masterConnectionMinimumIdleSize: 32
      masterConnectionPoolSize: 64
      readMode: "SLAVE"
      nodeAddresses:
        - "redis://xxxxx:7004"
        - "redis://xxxxx:7001"
        - "redis://xxxxx:7000"
      scanInterval: 1000
      keepAlive: true
      tcpNoDelay: true

```


Please See: https://github.com/redisson/redisson/blob/redisson-3.13.4/redisson/src/main/java/org/redisson/cluster/ClusterConnectionManager.java#L319  
``` java
                future.onComplete((r, ex) -> {
                            if (ex != null) {
                                log.error("unable to add slave for: " + partition.getMasterAddress()
                                                + " slot ranges: " + partition.getSlotRanges(), ex);
                            }
               }
````

Maybe set result in failure is a solution

``` java
                future.onComplete((r, ex) -> {
                            if (ex != null) {
                                log.error("unable to add slave for: " + partition.getMasterAddress()
                                                + " slot ranges: " + partition.getSlotRanges(), ex);
                                // set result in failure
                                result.tryFailure(ex);
                            }
               }
````


cc/ @mrniko

--- a/redisson/src/main/java/org/redisson/cluster/ClusterConnectionManager.java
+++ b/redisson/src/main/java/org/redisson/cluster/ClusterConnectionManager.java
@@ -313,11 +313,16 @@ public class ClusterConnectionManager extends MasterSlaveConnectionManager {
                 if (!config.checkSkipSlavesInit()) {
                     List<RFuture<Void>> fs = entry.initSlaveBalancer(partition.getFailedSlaveAddresses(), masterClient);
                     AtomicInteger counter = new AtomicInteger(fs.size());
+                    AtomicInteger errorCounter = new AtomicInteger(fs.size());
                     for (RFuture<Void> future : fs) {
                         future.onComplete((r, ex) -> {
                             if (ex != null) {
                                 log.error("unable to add slave for: " + partition.getMasterAddress()
                                                 + " slot ranges: " + partition.getSlotRanges(), ex);
+                                if (errorCounter.decrementAndGet() == 0) {
+                                    result.tryFailure(ex);
+                                    return;
+                                }
                             }
 
                             if (counter.decrementAndGet() == 0) {

